### DISCUSSION TITLE
怎么对服务端菜单返回的路由 进行更友好的重定向

### DISCUSSION THREAD

**Discussion by ssjshengjie** [umijs/umi#12023] [Q&A]
目前使用的mix 模式 菜单切割 没有办法 重定向  暂时 实现的方法是 
![image](https://github.com/umijs/umi/assets/56914431/f47ac35e-63f4-4a7b-9bd5-45a11c9fa587)
每次 切换路由的时候 去判断是不是第一级路由 如果是 并且有子菜单的情况下 重定向到子路由  看大佬说在patchClientRoutes  目前还是感觉不行 每次切换的时候 回去重新请求 menu 接口的同时  还是写死的重定向 不一定别的人有这个路由的权限

**Comment by fz6m** [comment]
如果使用了 patch 在运行时动态修改路由表，则权限应该事先最早就确定好，只展示保留有权限的。

一级菜单展示的内容你也可以自定义组件，在这个组件里看一下第一个用户有权限的子路由是哪个，可以使用 [`useAppData`](https://umijs.org/docs/api/api#useappdata) ，[`useSelectedRoutes`](https://umijs.org/docs/api/api#useselectedroutes) ，`useLocation` 等 api 来获取相关的信息辅助判断，或者你在 patch 的时候动态创建一个重定向组件给一级路由就行了，他大概的样子是：

```ts
export const patchClientRoutes = (...) => {
  push({
    element: <CustomRedirect to={/* 有权限的第一个子路由路径，当然在这个组件里动态判断也行 */} />
  })
}
```
